File: src/watchy/github.py
"""GitHub API client for fetching stargazers and followers."""

import json
import time
from typing import Iterator, Dict, Any, Optional
import requests


class GitHubClient:
    """Client for interacting with GitHub API."""

    def __init__(self, token: Optional[str] = None):
        self.session = requests.Session()
        if token:
            self.session.headers.update({"Authorization": f"token {token}"})
        self.session.headers.update({"Accept": "application/vnd.github.v3+json", "User-Agent": "watchy/0.1.0"})

    def _paginate(self, url: str, params: Optional[Dict[str, Any]] = None) -> Iterator[Dict[str, Any]]:
        """Paginate through GitHub API responses."""
        if params is None:
            params = {}
        params.setdefault("per_page", 100)

        while url:
            response = self.session.get(url, params=params)
            response.raise_for_status()

            # Handle rate limiting
            if response.status_code == 403 and "rate limit" in response.text.lower():
                reset_time = int(response.headers.get("X-RateLimit-Reset", 0))
                sleep_time = max(reset_time - int(time.time()) + 1, 60)
                print(f"Rate limited. Sleeping for {sleep_time} seconds...")
                time.sleep(sleep_time)
                continue

            data = response.json()
            for item in data:
                yield item

            # Get next page URL from Link header
            url = None
            if "Link" in response.headers:
                links = response.headers["Link"].split(",")
                for link in links:
                    if 'rel="next"' in link:
                        url = link.split(";")[0].strip("<> ")
                        break
            params = None  # Only use params for first request

    def get_stargazers(self, owner: str, repo: str) -> Iterator[Dict[str, Any]]:
        """Get stargazers for a repository."""
        url = f"https://api.github.com/repos/{owner}/{repo}/stargazers"
        params = {"per_page": 100}
        yield from self._paginate(url, params)

    def get_followers(self, user: str) -> Iterator[Dict[str, Any]]:
        """Get followers for a user or organization."""
        url = f"https://api.github.com/users/{user}/followers"
        params = {"per_page": 100}
        yield from self._paginate(url, params)

    def get_repositories(self, user: str) -> Iterator[Dict[str, Any]]:
        """Get repositories for a user or organization."""
        url = f"https://api.github.com/users/{user}/repos"
        params = {"per_page": 100, "type": "owner"}
        yield from self._paginate(url, params)

File: cfw/src/github.ts
/** Minimal GitHub REST client: token auth, Link-header pagination. */

const API = 'https://api.github.com'
const PER_PAGE = 100
const MAX_PAGES = 100

export interface Repo {
  full_name: string
  stargazers_count: number
}

export interface Stargazer {
  starred_at: string | null
  uid: number
  login: string
}

export interface Follower {
  uid: number
  login: string
}

export class GhError extends Error {
  constructor(readonly status: number, readonly url: string) {
    super(`GitHub ${status}: ${url}`)
  }
}

async function* paginate(token: string, path: string, accept?: string): AsyncGenerator<any> {
  let url: string | null = `${API}${path}${path.includes('?') ? '&' : '?'}per_page=${PER_PAGE}`
  let pages = 0
  while (url && pages++ < MAX_PAGES) {
    const resp: Response = await fetch(url, {
      headers: {
        authorization: `Bearer ${token}`,
        accept: accept ?? 'application/vnd.github.v3+json',
        'user-agent': 'watchy-cfw',
      },
    })
    if (!resp.ok) throw new GhError(resp.status, url)
    const items = (await resp.json()) as any[]
    yield* items
    url = null
    const link = resp.headers.get('link')
    if (link) {
      for (const part of link.split(',')) {
        if (part.includes('rel="next"')) {
          url = part.split(';')[0].trim().replace(/^<|>$/g, '')
          break
        }
      }
    }
  }
}

/** Public repos owned by a user/org (matches the py client: /users/{owner}/repos?type=owner). */
export async function listRepos(token: string, owner: string): Promise<Repo[]> {
  const repos: Repo[] = []
  for await (const r of paginate(token, `/users/${owner}/repos?type=owner`)) {
    if (r.private) continue
    repos.push({ full_name: r.full_name, stargazers_count: r.stargazers_count })
  }
  return repos
}

/** star+json accept variant → each entry carries `starred_at`. */
export async function listStargazers(token: string, repo: string): Promise<Stargazer[]> {
  const out: Stargazer[] = []
  for await (const s of paginate(token, `/repos/${repo}/stargazers`, 'application/vnd.github.star+json')) {
    out.push({ starred_at: s.starred_at ?? null, uid: s.user.id, login: s.user.login })
  }
  return out
}

export async function listFollowers(token: string, user: string): Promise<Follower[]> {
  const out: Follower[] = []
  for await (const f of paginate(token, `/users/${user}/followers`)) {
    out.push({ uid: f.id, login: f.login })
  }
  return out
}
